### ISSUE TITLE
Move core code from Bash to Python

### ISSUE THREAD

**Issue by Ornias1993** [jailmanager/jailman#202]
As discussed earlier, there a lot of benefids that make it worthwhile to move the core program to python.
However the plugins/blueprints will stay bash-based, to keep it accessable

**Comment by asmod3us** [comment]
I've cleaned up my experiment and published it at https://github.com/asmod3us/pailman. This is based on older jailman code pre-plugin switch, so it'll need some catching up. I have spend most of the time on a solid dev setup with unit testing, test coverage and config validation.

Next bigger decisions would be on CLI parsing (use a framework like click vs what's currently in the PoC)

**Comment by Ornias1993** [comment]
Thanks for the update, i'll make sure you get a branch on our repo for your work...

**Comment by Ornias1993** [comment]
I you a seperate out-of-tree branch, for which you should already have write access:
https://github.com/jailmanager/jailman/tree/pailman

**Comment by asmod3us** [comment]
Cool, thanks!

**Comment by Ornias1993** [comment]
The changes for the main code are not that big actually...

A TLDR:
- The 'iocage create' now calls a plugin-to-be-created instead of a json-with-pkgs
- The plugin INDEX repository gets checked instead of the blueprint folder, for all before-creation checks.
- Jail config object is moved to the root of config.yml (so no `jail:` section anymore)
- reqvars and vars are read from the plugin json (which is downloaded by iocage create automatically)

Most other fixes are either traefik related, renaming related (blueprints -> plugins) or just restructuring, the basics above are mostly what is relevant for your python work...

**Comment by asmod3us** [comment]
Looks like we mostly require renaming/moving some things around in the config parser and validator and refactor the reqvars/vars part to use json instead of yaml sources.

**Comment by Ornias1993** [comment]
I don't think you already added code for actually doing anything with the grabbed data, so the changes till this stage are rather minimal indeed:

Code wise this would translate in:
- Grabbing your req-vars from the json manifest instead of the config.yml of the blueprint
- changing the structure of the config.yml to have the jailnames at root level

I think we can drop "vars" from the python version and just load all variables decleared under a certain jail in config.yml, agreed?

**Comment by asmod3us** [comment]
That's correct, invoking the parts that do things are not yet implemented. I'll have to do more research to comment on the vars changes. What I wanted to do in the python version is to validate a plugin's configuration and inform the user if any required vars are not set or vars are set that a plugin does not support.

**Comment by Ornias1993** [comment]
> I'll have to do more research to comment on the vars changes

- When it comes to the config.yml for the plugin, it's basically just changed to a json and the config options for jailman are moved up one level under "jailman"

Example:
https://github.com/jailmanager/iocage-plugins/blob/master/sonarr.json

basically all content from the old `blueprints/sonarr/config.yml`
is available in this json under "jailman"

The reqvar checks are moved after jail creation and done like this currently:
`reqvars=$(jq -r '.jailman | .variables | .required | .[]' "${global_dataset_iocage}/jails/${jail}/${plugin}.json")`

> What I wanted to do in the python version is to validate a plugin's configuration and inform the user if any required vars are not set or vars are set that a plugin does not support.

In that case we could best indeed keep the "vars" configuration in the plugin manifest too... to inform users of vars being set that are not supported by the plugin! :)

**Comment by Ornias1993** [comment]
@asmod3us Just a short headsup:
I've spend last months on looking into the new (linux based) truenas SCALE system and it seems we might be able to utilise the UI, to complety cut the "user needs to write a config file and use CLI" issues...

Currently working on porting everything we have here to the new system, so it is ready at launch:
https://github.com/truecharts/truecharts

That being said: It's not clear if we would be able to have our own TrueNAS SCALE app catalog available without pointing people to a script. So you might also want to upvote this suggestion (which gets serieus feedback when 10 people upvote it):
https://jira.ixsystems.com/browse/NAS-108842

**Comment by jsegaert** [comment]
Thanks for trailblazing this!  I upvoted the JIRA suggestion!  👍

**Comment by Ornias1993** [comment]
@jsegaert I've made a complete writedown of the reasoning behind moving to another solution here:
https://www.truenas.com/community/resources/jailman-automated-management-of-pseudo-custom-freenas-jails.135/updates#resource-update-382

In case you missed it :)

**Comment by Ornias1993** [comment]
past month-and-a-half, App development for TrueNAS SCALE has taken a big increase. A lot of apps are already included in either TrueCharts or the official TrueNAS SCALE App repository.

How is this related to this issue?
The middleware for the SCALE Apps is already writhen in Python, is pretty solid and is opensource. In that respect TrueNAS SCALE makes sure I/We can focus on getting awesome Apps ported to TrueNAS, instead of writhing our own middleware-middleware :)

This means I invite everyone here (hint @colemamd @asmod3us @disconn3ct and @thorsteneb ) to join me at https://github.com/truecharts/truecharts as I think that project has a lot more future than this one. (for those of you with write/review access, expect those being ported already. if not: send me an email)
